feat(ui): add 'statuscolumn' option (#20621)
neovim/neovim: runtime/doc/options.txt

	    |nobody         |  3 nobody     |  0 nobody     |3   nobody
	    |there          |  4 there      |  1 there      |  1 there

				*'numbercolumn'* *'nuc'*
'numbercolumn' 'nuc'	string	(default: empty)
			local to window
	EXPERIMENTAL
	When this option is not empty, it determines the content of the number
	column. The format of this option is like that of 'statusline'. The
	variables *v:redraw_lnum* and *v:redraw_relnum* will be set to the to be
	printed line number and relative line number respectively. These can be
	evaluated in an expression inside a 'statusline' %{} block.

	Some of the items from the 'statusline' format are different
	'numbercolumn':

> 	%l	v:redraw_lnum, can be used inside %{} block

Reviewer: This is wrong `%l` cannot be used in `%{}`. Make sure to document `%l` and `v:redraw_lnum` clearly.

Author: It's not wrong in this PR. I have altered the `%{}` eval block to substitutie `v:redraw_lnum` and `v:redraw_relnum` for `%l` and `%r` respectively: https://github.com/luukvbaal/neovim/blob/93a9277b76b76e0a0028ebf5aa5136b6418ce1e7/src/nvim/statusline.c#L1152-L1158 (see https://github.com/neovim/neovim/pull/20621#discussion_r998503306)

This allows you to write e.g. `set nuc=%{%r?%r:%l}` instead of `set nuc=%{v:redraw_relnum?v:redraw_relnum:v:redraw_lnum}`. Might not be worth it.

Reviewer: Not a fan personally. I'd reserve `%{}` blocks for viml expressions.

Reviewer 2: > Not a fan personally. I'd reserve %{} blocks for viml expressions.

Agreed. We shouldn't change the behavior of `%{}`.

> This allows you to write e.g. `set nuc=%{%r?%r:%l}` instead of `set nuc=%{v:redraw_relnum?v:redraw_relnum:v:redraw_lnum}`

There's already a syntax for re-evaluating the expression result - `%{%%}`. Wouldn't something like `%{%'%r?%r:%l'%}` acheive the same result? 🤔

Author: No that wouldn't work. The number needs to be substituted before the viml expression is evaluated. But this change has felt dubious to me from the start as well. I'll revert it.

	%r	v:redraw_relnum, can be used inside %{} block
	%s	Sign column for v:redraw_lnum

> 	%C	Fold column for v:redraw_lnum

Reviewer: The sign/fold column should be separated out of `v:redraw_lnum` which should just be an integer.

Author: It is. What I'm saying here is that `%s` and `%C` get the sign/fold column content for the currently drawn line, which is equal to `v:redraw_lnum`. Perhaps not too clear.

Reviewer: Ok, then this just needs some minor rewording.

	    |nobody         |  3 nobody     |  0 nobody     |3   nobody
	    |there          |  4 there      |  1 there      |  1 there

				*'numberformat'* *'nuf'*
'numberformat' 'nuf'	string	(default: empty)
			local to window
	When this option is not empty, it determines the content of the number
	column. The format of this option is like that of 'statusline'. The

> 	variables *g:nuflnum* and *g:nufrnum* will be set to the to be printed line

Reviewer 3: can/should they be `v:` vars? Similar to `v:event`, maybe a common container like `v:redraw` is a good pattern. So these would be `v:redraw.lnum` and `v:redraw.relnum`.

Related to performance, accessing vimscript dicts (`g:`, `v:`) is slow, but we might be able to optimize `v:` vars because they are readonly.

Author: Sounds reasonable, also left another comment regarding performance that would require adding new 'statusline' expression items.

Author: With the below benchmark by running `time nvim --clean -S test.vim`(commenting out the relevant line):
```vim
set number nowrap
set numbercolumn=%l
"set numbercolumn=%{v:redraw_lnum}
"set numberformat=%{nuflnum}

norm ia
norm yy9999P
let i = 0
while i < 9999
	let i += 1
	norm j
  redraw!
endwhile
quit!
```
There is no way of telling(with this rudimentary test) whether using a read-only `v:` scope variable is actually faster than a `g:` scope variable (currently, because you say it might still be able to be optimized). The lowest runtime alternates between the g:/v: builds in my testing.

The `set numbercolumn=%l` variant, without an `%{}` eval block, **is** measurably faster. Presumably the overhead in setting up the eval environment(not sure that's proper jargon) is the determining factor here, making differences(if any) in variable scope insignificant.

So there might be merit in adding new `'statusline`' expression items specifically for `'numbercolumn'`. It feels kind of iffy since they wouldn't be useful for the other statusline-format options but we might even go so far as to provide things like hex/thousands separator (relative)line number items as well. That way most users wouldn't need an eval block in their `'numbercolumn'` I think, but we can still set variables for those that do. Different colors for for odd/even line numbers is one mentioned use case that comes to mind where evaluating the line number would be necessary.

> > We could also add new format expression items containing the to be printed (relative)number.
 
> Rather not. Let's optimize `v:` access instead (if it's not already optimized).

Considering the above, would you still rather not? Care to elaborate?

Reviewer 3: Let's stick with `v:` for now. Could also overload `%l` as you discuss in https://github.com/neovim/neovim/pull/20621#discussion_r994894476

	    |nobody         |  3 nobody     |  0 nobody     |3   nobody
	    |there          |  4 there      |  1 there      |  1 there

				*'numbercolumn'* *'nuc'*
'numbercolumn' 'nuc'	string	(default: empty)
			local to window
	EXPERIMENTAL
	When this option is not empty, it determines the content of the number
	column. The format of this option is like that of 'statusline'. The

> 	variables *v:redraw_lnum* and *v:redraw_relnum* will be set to the to be printed line

Reviewer: We can re-use `%l` instead of introducing a new variable. For relative line number we can either use a free letter (no obvious choices) or just overload and existing one (maybe `%r`) since most of them aren't applicable here.

Author: Consider that `%l` would write the cursor line number for each line. That would have to be overloaded too for it to return the actual number line to be drawn. But yeah overloading seems reasonable since most of seem irrelevant for the number column.

Reviewer: Hmm, just realized `%l` can't be used within `%{ }` so a `v:` variable will be necessary.

Author: Yes, that is an unfortunate limitation but probably hard to overcome. Unless if we somehow pre-parse the %{} block for restricted characters to substitute the line number. Not sure that's feasible or worth it.

Author: Again, I think there could still be merit in adding/overloading stl items for 'cursorcolumn' for simple cases like `'set nuf=%=%l'`. But yeah I think the variables need to stay regardless.

	In case of buffer changing commands the cursor is placed at the column
	where it was the last time the buffer was edited.

				*'statuscolumn'* *'stc'*
'statuscolumn' 'stc'	string	(default: empty)
			local to window
	EXPERIMENTAL

> 	When this option is not empty, it determines the content of the area to

Reviewer: ```suggestion
	When non-empty, determines the content of the area to
```

Author: Sure, the same wording is used for `iconstring`, `indentexpr`, `rulerformat` and `titlestring` btw, that's why I used it. `statusline` on the other hand reads:

    	When non-empty, this option determines the content of the status line.

I guess I'll word it similar to that.

Reviewer: :+1:

	the side of a window, normally containing the fold, sign and number
	columns. The format of this option is like that of 'statusline'.

	Some of the items from the 'statusline' format are different for
	'statuscolumn':

	%l	line number of currently drawn line
	%r	relative line number of currently drawn line
	%s	sign column for currently drawn line
	%C	fold column for currently drawn line

	To draw the sign and fold columns, they must be included in
	'statuscolumn'.

	The |v:lnum|   variable holds the line number to be drawn.
	The |v:relnum| variable holds the relative line number to be drawn.
	The |v:wrap|   variable holds true for the wrapped part of a line.

> 	Examples: >

Reviewer: ```suggestion
	Examples: >vim
```
